### PR TITLE
GH-109 '/autolink list' enable searching patterns and templates

### PR DESCRIPTION
**Pull request by GrigalashviliT**
#### Summary
Adds possibilities to list links according to Template or Pattern values 

#### Ticket Link
  Fixes https://github.com/mattermost/mattermost-plugin-autolink/issues/109

### CODE REVIEW

File: server/autolinkplugin/command.go [hunk 1]
```diff
@@ -12,13 +12,16 @@ import (
 	"github.com/mattermost/mattermost-plugin-autolink/server/autolink"
 )
 
+const template, pattern = "Template", "Pattern"
```
Comment (levb) [inline server/autolinkplugin/command.go:15]: 2/5 if we parameterize command names we should do it for all commands, and use more intention-revealing names, perhaps `optTemplate` etc.

@iomodo we should file a ticket to convert autolink to autocomplete!

File: server/autolinkplugin/command.go [hunk 2]
```diff
@@ -85,7 +88,15 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, commandArgs *model.CommandArg
 }
 
 func executeList(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
-	links, refs, err := parseLinkRef(p, false, args...)
+	var links []autolink.Autolink
+	var refs []int
+	var err error
+
+	if len(args) > 0 && (args[0] == template || args[0] == pattern) {
+		links, refs, err = parseLinkRefByTemplateOrPattern(p, header, args...)
```
Comment (levb) [inline server/autolinkplugin/command.go:96]: 0/5 rename the `parseLink...` functions to `searchLink...` to be more intention-revealing?

File: server/autolinkplugin/command.go [hunk 3]
```diff
@@ -329,6 +340,31 @@ func parseLinkRef(p *Plugin, requireUnique bool, args ...string) ([]autolink.Aut
 	return links, found, nil
 }
 
+func parseLinkRefByTemplateOrPattern(p *Plugin, header *model.CommandArgs, args ...string) ([]autolink.Autolink, []int, error) {
+	links := p.getConfig().Sorted().Links
+	if len(args) == 1 {
+		return links, nil, nil
+	}
+
+	restOfCommand := header.Command[10:]
```
Comment (levb) [inline server/autolinkplugin/command.go:349]: 1/5 `10` must be a length of something... better expressed as len(x)

File: server/autolinkplugin/command.go [hunk 4]
```diff
@@ -12,7 +12,8 @@ import (
 	"github.com/mattermost/mattermost-plugin-autolink/server/autolink"
 )
 
-const template, pattern = "Template", "Pattern"
+const optName, optTemplate, optPattern, optScope, optDisabled, optDisableNonWordPrefix, optDisableNonWordSuffix, optWordMatch = "Name", "Template", "Pattern", "Scope", "Disabled", "DisableNonWordPrefix", "DisableNonWordSuffix", "WordMatch"
```
Comment (levb) [inline server/autolinkplugin/command.go:15]: 1/5 format as 1/line for readability/typography

File: server/autolinkplugin/command.go [hunk 5]
```diff
@@ -12,7 +12,8 @@ import (
...
+const command = "/autolink"
```
Comment (levb) [inline server/autolinkplugin/command.go:16]: 0/5 make it a bit more specific, intention-revealing; maybe `autolinkCommand` or `cmdAutolink`?

File: server/autolinkplugin/command.go (hunk 1 above)
Comment (iomodo) [reply server/autolinkplugin/command.go]: It already has been filed https://github.com/mattermost/mattermost-plugin-autolink/issues/118

File: server/autolinkplugin/command.go [hunk 6]
```diff
@@ -329,6 +350,31 @@ func parseLinkRef(p *Plugin, requireUnique bool, args ...string) ([]autolink.Aut
...
+	if len(args) == 1 {
+		return links, nil, nil
+	}
+
+	restOfCommand := header.Command[len(autolinkCommand):]
+	restOfCommand = restOfCommand[strings.Index(restOfCommand, args[0])+len(args[0]):]
+	value := strings.TrimSpace(restOfCommand)
+
+	found := []int{}
+	for i, l := range links {
```
Comment (jfrerich) [inline server/autolinkplugin/command.go:364]: Outside the scope of this PR, but after just a quick look, I don't see why we need to return the array of indexes. Couldn't the caller just iterate over the returned links? @levb

Comment (levb) [reply server/autolinkplugin/command.go:364]: I can't remember off the top of my head, but does it have to do with the indices listed with `/autolink list`?

**Comment by DHaussermann** [comment]
@GrigalashviliT thank you for this improvement! Sorry for the delay in testing this. I've taken a look and it seems to be working as expected.

I have only one minor point to bring up. Would it take much effort to make the `value` case  insensitive. For example `/autolink list Template  scrum` would work the same as `/autolink list Template  SCRUM`.  Since something like Jira issue keys would have upper case letters. So, if it's simple to make the value case insensitive, it would improve the usability even more.

**Comment by DHaussermann** [comment]
Thanks @GrigalashviliT for this enhancement!

I see there is still a `Docs/Needed` label on this. The help command has been updated. Should we also update the table in the read-me with the new list command?
@levb or @iomodo, I'm not certain who does read-me updates or who removes this label.

**Pull request merged**